pkg/kv/kvserver/replica_raft_quiesce.go

```diff
@@ -337,15 +339,34 @@ func shouldReplicaQuiesce(
 		}
 		return nil, nil, false
 	}
-	// Only quiesce if this replica is the leaseholder as well;
-	// otherwise the replica which is the valid leaseholder may have
-	// pending commands which it's waiting on this leader to propose.
-	if !q.ownsValidLeaseRLocked(ctx, now) {
+
+	// Don't quiesce if there is a current leaseholder elsewhere. Otherwise, the
+	// leaseholder may have pending commands which it's waiting on this leader to
+	// propose.
+	//
+	// We allow quiescing with an expired lease (both expiration-based and
+	// epoch-based), since leases are not always eagerly renewed. Noone else can
+	// have acquired the lease at this point, because this replica is the leader
```

Comment:
That's not strictly true: the replica believes it's the leader but there could be another one. Correspondingly, this replica's view of its followers might be incorrect.

I still think we should do this - the effects of ~permanently leaving potentially lots of ranges unquiesced should their leases run out is problematic, could even be a stability risk.

What are the effects of accidentally quiescing? The other live replicas will ignore the quiesce since the leader/term doesn't match. So all that would occur here is that an out-of-touch leader would quiesce. It would promptly get woken up again, or, in the worst case, catch an election timeout when it's needed again to make progress (but I think we'll eagerly campaign in this case anyway). These cases should be exceedingly rare.

Reply:
Good catch, I verified this and agree with your take. Updated the comment.